File: backend/app/api/endpoints/applications.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy.orm import Session

import app.services.applications as applications_service
from app.api.deps import get_current_user, get_db
from app.models.users import User
from app.schemas.applications import (
    ApplicationCreate,
    ApplicationUpdate,
    ApplicationResponse,
    DashboardResponse,
)


router = APIRouter()


@router.post("/")
def create_application(
    application: ApplicationCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    return applications_service.create_application(db, current_user, application)


@router.get("/")
def get_applications(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    offset: int = 0,
    limit: int = 10,
) -> list[ApplicationResponse]:
    return applications_service.get_applications(db, current_user.id, offset, limit)


@router.get("/dashboard")
def get_dashboard(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> DashboardResponse:
    return applications_service.get_dashboard(db, current_user.id)


@router.get("/{application_id}")
def get_application(
    application_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> ApplicationResponse:
    application = applications_service.get_application(
        db, application_id, current_user.id
    )
    if application is None:
        raise HTTPException(status_code=404, detail="Candidatura não encontrada.")
    return application


@router.put("/{application_id}")
def update_application(
    application_id: int,
    application: ApplicationUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    result = applications_service.update_application(
        db, application_id, current_user.id, application
    )
    if result:
        return Response(status_code=204)
    return Response(status_code=404)


@router.delete("/{application_id}")
def delete_application(
    application_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    result = applications_service.delete_application(
        db, application_id, current_user.id
    )
    if result:
        return Response(status_code=204)
    return Response(status_code=404)

File: frontend/src/services/applications.ts
import type { ApplicationCreate, ApplicationUpdate } from "../types/application"
import { axiosClient } from "./client"

export async function getApplications(
    offset: number = 0,
    limit: number = 10
) {
    const response = await axiosClient.get(`/applications/?offset=${offset}&limit=${limit}`)
    return response.data
}

export async function createApplication(
    application: ApplicationCreate
) {
    const response = await axiosClient.post('/applications/', application)
    if (response.status === 201) {
        return response.data
    }
    throw new Error('Não foi possível criar a candidatura.')
}

export async function getApplication(
    id: number
) {
    const response = await axiosClient.get(`/applications/${id}`)
    if (response.status === 200) {
        return response.data
    }
    throw new Error('Não foi possível carregar a candidatura.')
}

export async function updateApplication(
    id: number,
    application: ApplicationUpdate
) {
    const response = await axiosClient.put(`/applications/${id}`, application)
    if (response.status === 204) {
        return
    }
    throw new Error('Não foi possível atualizar a candidatura.')
}


export async function deleteApplication(
    id: number
) {
    const response = await axiosClient.delete(`/applications/${id}`)
    if (response.status === 204) {
        return
    }
    throw new Error('Não foi possível excluir a candidatura.')
}


export async function getDashboard() {
    const response = await axiosClient.get('/applications/dashboard')
    if (response.status === 200) {
        return response.data
    }
    throw new Error('Não foi possível carregar o dashboard.');
}
